File: backend/__SLUG__/src/domains/notifications/service.go
package notifications

import (
	"context"
	"fmt"

	"firebase.google.com/go/v4/messaging"
	"go.uber.org/zap"
	"go_backend/src/logger"
)

// FCMSender is satisfied by *messaging.Client
type FCMSender interface {
	Send(ctx context.Context, msg *messaging.Message) (string, error)
}

type notifDatasource interface {
	UpsertDeviceToken(ctx context.Context, t *DeviceToken) (int, error)
	DeleteDeviceToken(ctx context.Context, userID, token string) error
	DeleteDeviceTokenByValue(ctx context.Context, token string) error
	GetTokensByUserID(ctx context.Context, userID string) ([]DeviceToken, error)
	InsertNotificationLog(ctx context.Context, log *NotificationLog) (int, error)
	UpdateNotificationLogStatus(ctx context.Context, id int, status, fcmMessageID, errMsg string) error
	GetNotificationLogs(ctx context.Context, filter LogQueryFilter) ([]NotificationLog, int, error)
}

type Service struct {
	ds     notifDatasource
	sender FCMSender
}

func NewService(ds notifDatasource, sender FCMSender) *Service {
	return &Service{ds: ds, sender: sender}
}

func (s *Service) RegisterToken(ctx context.Context, userID string, req *RegisterTokenRequest) (int, error) {
	return s.ds.UpsertDeviceToken(ctx, &DeviceToken{
		UserID:          userID,
		Token:           req.Token,
		Platform:        req.Platform,
		PlatformVersion: req.PlatformVersion,
		AppVersion:      req.AppVersion,
		DeviceID:        req.DeviceID,
	})
}

func (s *Service) DeleteToken(ctx context.Context, userID, token string) error {
	return s.ds.DeleteDeviceToken(ctx, userID, token)
}

func (s *Service) Send(ctx context.Context, req *SendNotificationRequest) (*SendNotificationResponse, error) {
	if req.Type == "alert" && (req.Title == "" || req.Body == "") {
		return nil, fmt.Errorf("title and body required for alert type")
	}
	if len(req.UserIDs) == 0 {
		return nil, fmt.Errorf("user_ids must not be empty")
	}

	log := logger.LoggerFromContext(ctx)
	log.Info("send notification request",
		zap.Strings("user_ids", req.UserIDs),
		zap.String("type", req.Type),
		zap.String("title", req.Title),
	)

	var logIDs []int
	queued := 0

	for _, userID := range req.UserIDs {
		tokens, err := s.ds.GetTokensByUserID(ctx, userID)
		if err != nil {
			return nil, fmt.Errorf("get tokens for user %s: %w", userID, err)
		}

		log.Info("resolved device tokens",
			zap.String("user_id", userID),
			zap.Int("token_count", len(tokens)),
		)

		for _, t := range tokens {
			log.Info("sending FCM message",
				zap.String("user_id", userID),
				zap.String("platform", t.Platform),
				zap.String("device_id", t.DeviceID),
				zap.String("token_prefix", truncate(t.Token, 16)),
			)

			data := buildDataPayload(req)
			logID, err := s.ds.InsertNotificationLog(ctx, &NotificationLog{
				UserID:           userID,
				DeviceToken:      t.Token,
				NotificationType: req.Type,
				Title:            req.Title,
				Body:             req.Body,
				DataPayload:      toStringMap(data),
			})
			if err != nil {
				return nil, fmt.Errorf("insert log: %w", err)
			}
			logIDs = append(logIDs, logID)
			queued++

			msg := &messaging.Message{
				Token: t.Token,
				Data:  data,
				Android: &messaging.AndroidConfig{
					Priority: "high",
				},
				APNS: &messaging.APNSConfig{
					Headers: map[string]string{
						"apns-push-type": "background",
						"apns-priority":  "5",
					},
					Payload: &messaging.APNSPayload{
						Aps: &messaging.Aps{
							ContentAvailable: true,
						},
					},
				},
			}

			messageID, fcmErr := s.sender.Send(ctx, msg)
			status, errMsg := "sent", ""
			if fcmErr != nil {
				status = "failed"
				errMsg = fcmErr.Error()
				log.Error("FCM send failed",
					zap.String("user_id", userID),
					zap.String("platform", t.Platform),
					zap.String("device_id", t.DeviceID),
					zap.String("type", req.Type),
					zap.Error(fcmErr),
				)
			} else {
				log.Info("FCM send ok",
					zap.String("user_id", userID),
					zap.String("platform", t.Platform),
					zap.String("device_id", t.DeviceID),
					zap.String("type", req.Type),
					zap.String("message_id", messageID),
				)
			}
			_ = s.ds.UpdateNotificationLogStatus(ctx, logID, status, messageID, errMsg)

			if fcmErr != nil && (messaging.IsRegistrationTokenNotRegistered(fcmErr) || messaging.IsInvalidArgument(fcmErr)) {
				log.Info("removing stale token",
					zap.String("user_id", userID),
					zap.String("device_id", t.DeviceID),
				)
				_ = s.ds.DeleteDeviceTokenByValue(ctx, t.Token)
			}
		}
	}

	log.Info("send notification complete", zap.Int("queued", queued))
	return &SendNotificationResponse{Queued: queued, LogIDs: logIDs}, nil
}

func (s *Service) GetLogs(ctx context.Context, requesterUserID string, isAdmin bool, filter LogQueryFilter) (*LogQueryResponse, error) {
	if !isAdmin {
		filter.UserID = requesterUserID
	}
	logs, total, err := s.ds.GetNotificationLogs(ctx, filter)
	if err != nil {
		return nil, err
	}
	return &LogQueryResponse{Total: total, Results: logs}, nil
}

func buildDataPayload(req *SendNotificationRequest) map[string]string {
	data := map[string]string{
		"type": req.Type,
	}
	if req.Type == "alert" {
		data["title"] = req.Title
		data["body"] = req.Body
	}
	for k, v := range req.Data {
		data[k] = v
	}
	return data
}

func truncate(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n] + "..."
}

func toStringMap(m map[string]string) map[string]interface{} {
	out := make(map[string]interface{}, len(m))
	for k, v := range m {
		out[k] = v
	}
	return out
}

File: backend/__SLUG___node/src/domains/notifications/service.ts
import type { Messaging, Message } from "firebase-admin/messaging";
import { loggerFromContext } from "../../logger/logger";
import type { Datasource } from "./datasource";
import type {
  LogQueryFilter,
  LogQueryResponse,
  RegisterTokenRequest,
  SendNotificationRequest,
  SendNotificationResponse,
} from "./models";

function truncate(s: string, n: number): string {
  return s.length <= n ? s : `${s.slice(0, n)}...`;
}

function buildDataPayload(req: SendNotificationRequest): Record<string, string> {
  const data: Record<string, string> = { type: req.type };
  if (req.type === "alert") {
    data.title = req.title;
    data.body = req.body;
  }
  Object.assign(data, req.data ?? {});
  return data;
}

export class Service {
  constructor(
    private readonly ds: Datasource,
    private readonly messaging: Messaging,
  ) {}

  registerToken(userId: string, req: RegisterTokenRequest): Promise<number> {
    return this.ds.upsertDeviceToken({
      userId,
      token: req.token,
      platform: req.platform,
      platformVersion: req.platform_version,
      appVersion: req.app_version,
      deviceId: req.device_id,
    });
  }

  deleteToken(userId: string, token: string): Promise<void> {
    return this.ds.deleteDeviceToken(userId, token);
  }

  async send(req: SendNotificationRequest): Promise<SendNotificationResponse> {
    if (req.type === "alert" && (!req.title || !req.body)) {
      throw new Error("title and body required for alert type");
    }
    if (!req.user_ids || req.user_ids.length === 0) {
      throw new Error("user_ids must not be empty");
    }

    const log = loggerFromContext();
    log.info({ user_ids: req.user_ids, type: req.type, title: req.title }, "send notification request");

    const logIds: number[] = [];
    let queued = 0;

    for (const userId of req.user_ids) {
      const tokens = await this.ds.getTokensByUserId(userId);
      log.info({ user_id: userId, token_count: tokens.length }, "resolved device tokens");

      for (const t of tokens) {
        log.info(
          { user_id: userId, platform: t.platform, device_id: t.deviceId, token_prefix: truncate(t.token, 16) },
          "sending FCM message",
        );

        const data = buildDataPayload(req);
        const logId = await this.ds.insertNotificationLog({
          userId,
          deviceToken: t.token,
          notificationType: req.type,
          title: req.title,
          body: req.body,
          dataPayload: data,
        });
        logIds.push(logId);
        queued++;

        const message: Message = {
          token: t.token,
          data,
          android: { priority: "high" },
          apns: {
            headers: { "apns-push-type": "background", "apns-priority": "5" },
            payload: { aps: { contentAvailable: true } },
          },
        };

        let status = "sent";
        let errMsg = "";
        let messageId = "";
        try {
          messageId = await this.messaging.send(message);
          log.info(
            { user_id: userId, platform: t.platform, device_id: t.deviceId, type: req.type, message_id: messageId },
            "FCM send ok",
          );
        } catch (err) {
          status = "failed";
          errMsg = (err as Error).message;
          log.error(
            { user_id: userId, platform: t.platform, device_id: t.deviceId, type: req.type, err },
            "FCM send failed",
          );
        }
        await this.ds.updateNotificationLogStatus(logId, status, messageId, errMsg);

        // firebase-admin surfaces these as messaging/registration-token-not-registered
        // and messaging/invalid-argument on the thrown FirebaseMessagingError.
        if (
          status === "failed" &&
          (errMsg.includes("registration-token-not-registered") || errMsg.includes("invalid-argument"))
        ) {
          log.info({ user_id: userId, device_id: t.deviceId }, "removing stale token");
          await this.ds.deleteDeviceTokenByValue(t.token);
        }
      }
    }

    log.info({ queued }, "send notification complete");
    return { queued, log_ids: logIds };
  }

  async getLogs(requesterUserId: string, isAdmin: boolean, filter: LogQueryFilter): Promise<LogQueryResponse> {
    const effectiveFilter = isAdmin ? filter : { ...filter, userId: requesterUserId };
    const { logs, total } = await this.ds.getNotificationLogs(effectiveFilter);
    return { total, results: logs };
  }
}
